### PR TITLE
Stops AI & Borg upload consoles working on other zlevels

### PR DESCRIPTION
**Pull request by Kyep**
AI upload, and borg upload, consoles now only work on the station zlevel.

🆑
tweak: AI & Borg upload consoles now have to be on the station zlevel to work.
/🆑

### CODE REVIEW

**Comment by Anasari** [comment]
+1 as this make AI sabotage more counterable instead of a risk-free endeavour that must end with the AI being destroyed.

**Comment by Anoonki** [comment]
> +1 as this make AI sabotage more counterable instead of a risk-free endeavour that must end with the AI being destroyed.

Yesplease.
I'm sick of needing to card the AI or kill it outright to keep it from actively killing everyone.
Especially when *certain* unnamed individuals will do it from a location they know full well that they will *never* be caught.

**Comment by DesolateG** [comment]
+1

**Comment by Fraility1** [comment]
I am happy with it working in certain z-levels. like those with telecoms like Research outpost, mining and engineering outpost places where it could be found if you were trying to look for it. but things like the abandoned station and stuff like that i am not okay with.

**Comment by Kyep** [comment]
Realistically, if someone constructs an AI upload in a decent place on the eng sat, sci outpost, etc, it probably won't be found. And the AI will get continually re-lawed with bad laws for the rest of the round. Short of searching the whole station AND the outposts AND wherever else it can be put, its uncounterable. They can just keep resetting the laws.

**Comment by davidchan13** [comment]
Can you add the Robotics Control console to this (Borg lockdown/destruct.)

But yeah :+1:  this is something that should have been done years ago, remotely changing laws from another z level is ridiculous. I'd even take it a step farther and notify the AI/Borgs if an upload is constructed.

**Comment by SamHPurp** [comment]
@davidchan13 I'm not a fan of the idea of notification on construction of an upload.

But I fully support this PR... this makes subversion more counterable... and when I'm subverted as an AI... less chance I'll get toxin bombed and never repaired.

**Comment by MarsM0nd** [comment]
Why force the uploads to be on station, if it's possible that the synthetics, especially the borgs, might be off-station. Might make a bit more sense to have them work like crew monitors, you can upload laws to everything on the same z-level.

**Comment by Kyep** [comment]
@MarsM0nd That makes no sense. We don't have player-controlled AIs or borgs that live off station. Even if we did, we'd probably want them controlled via AI upload, like existing AIs. Changing the requirement to "same z-level" to support the non-existent case of AIs/borgs that live off station, when this isn't even necessarily desirable, is not an improvement.

**Comment by DesolateG** [comment]
The asteroid and engineering z levels are also extremely capable of being cheesed with an upload computer no one will find.

Construction station and the asteroid is huge.

**Comment by Taaketa** [comment]
I always thought this was the case and then discovered it wasn't. Last time I did it. I just uploaded the rules of Fight Club to the AI.

**Comment by Necaladun** [comment]
It makes sense that they could only affect things on the same Z level.

And stops it being too ez. All for this.

**Comment by SheepFighter** [comment]
While I agree that being able to upload from the drug lab and the derelict is cheesy, being able to do so from an outpost seems perfectly reasonable. Preventing antags from being able to do so only pigeonholes them more. 

Methinks security can afford to take the time to search the outposts. Why not allow the outposts while forbidding any other off-station areas?

**Comment by SamHPurp** [comment]
I subvert AIs all the time. You don't need to be off station. Its still really easy to do it on station.

Allowing outposts as well isn't a good idea.

**Comment by Fox-McCloud** [comment]
This only addresses the laptop version; not the main one.

Also, as others mentioned, this should target borgs/AI's on the same z-level as the console itself; if they're on the same level, the check passes and it goes through; if they're not on the same z-level, it'll fail.

**Comment by Kyep** [comment]
@Fox-McCloud Changed

File: code/modules/computer3/computers/law.dm [hunk 1]
```diff
@@ -24,7 +24,7 @@
 
 
 	attackby(obj/item/weapon/aiModule/module as obj, mob/user as mob, params)
-		if(!is_station_level(user.z))
+		if(is_away_level(user.z))
```
Comment (Fox-McCloud) [inline code/modules/computer3/computers/law.dm]: You're going to have to check for if the person is on the same z-level as the `current`'s turf z-level.

File: code/__HELPERS/unsorted.dm [hunk 2]
```diff
@@ -297,6 +297,18 @@ Turf and target are seperate in case you want to teleport some distance from a t
 		. += A
 	return .
 
+/proc/active_ais_samezlevel(/mob/M)
```
Comment (Fox-McCloud) [inline code/__HELPERS/unsorted.dm]: This isn't necessary.

**Comment by Kyep** [comment]
@Fox-McCloud changed

**Comment by ktoma36** [comment]
What about mining cyborgs that would spend most of their time on the asteroid? It would not make sense for them to miss a law set change. Perhaps having a relay added to the asteroid and other relevant Z-levels that would be synced with the station's upload could be implemented. If the relay was disabled the borgs on that Z-level would not receive law changes. These relays could also be constructed to cover other Z-Levels and perhaps even the gateway.

**Comment by Fox-McCloud** [comment]
>What about mining cyborgs that would spend most of their time on the asteroid? It would not make sense for them to miss a law set change

Then 'jack the AI.

File: code/modules/computer3/computers/law.dm [hunk 3]
```diff
@@ -60,6 +61,10 @@
 
 	attackby(obj/item/weapon/aiModule/module as obj, mob/user as mob, params)
 		if(istype(module, /obj/item/weapon/aiModule))
+			var/turf/T = get_turf(current)
+			if(T.z != z)
```
Comment (Crazylemon64) [inline code/modules/computer3/computers/law.dm]: Could this be changed to `atoms_share_level(T, src)`, as with above, to facilitate expansion of the space management system?

**Pull request merged**